File: day16_part2_2017.php

<?php

$input = file_get_contents("input.txt");
$moves = explode(",", $input);

$programs = str_split("abcdefghijklmnop");
$initial = implode("", $programs);
$cycleLen = 0;

for ($i = 0; $i < 1000000000; $i++) {
    foreach ($moves as $move) {
        switch ($move[0]) {
            case 's':
                $x = (int)substr($move, 1);
                spin($programs, $x);
                break;
            case 'x':
                $positions = explode("/", substr($move, 1));
                $A = (int)$positions[0];
                $B = (int)$positions[1];
                exchange($programs, $A, $B);
                break;
            case 'p':
                $positions = explode("/", substr($move, 1));
                $A = $positions[0];
                $B = $positions[1];
                partner($programs, $A, $B);
                break;
        }
    }

    if (implode("", $programs) == $initial) {
        $cycleLen = $i + 1;
        break;
    }
}

$programs = str_split($initial);

for ($i = 0; $i < 1000000000 % $cycleLen; $i++) {
    foreach ($moves as $move) {
        switch ($move[0]) {
            case 's':
                $x = (int)substr($move, 1);
                spin($programs, $x);
                break;
            case 'x':
                $positions = explode("/", substr($move, 1));
                $A = (int)$positions[0];
                $B = (int)$positions[1];
                exchange($programs, $A, $B);
                break;
            case 'p':
                $positions = explode("/", substr($move, 1));
                $A = $positions[0];
                $B = $positions[1];
                partner($programs, $A, $B);
                break;
        }
    }
}

echo implode("", $programs) . PHP_EOL;

function spin(&$programs, $x) {
    $n = count($programs);
    $temp = $programs;

    for ($i = 0; $i < $n; $i++) {
        $programs[($i + $x) % $n] = $temp[$i];
    }
}

function exchange(&$programs, $A, $B) {
    list($programs[$A], $programs[$B]) = array($programs[$B], $programs[$A]);
}

function partner(&$programs, $A, $B) {
    $indexA = array_search($A, $programs);
    $indexB = array_search($B, $programs);
    exchange($programs, $indexA, $indexB);
}
?>

File: day16_part2_2017.py

def spin(s, p):
    return p[-s:] + p[:-s]

def exchange(a, b, p):
    p[a], p[b] = p[b], p[a]
    return p

def partner(a, b, p):
    a_idx = p.index(a)
    b_idx = p.index(b)
    p[a_idx], p[b_idx] = p[b_idx], p[a_idx]
    return p

def dance(moves, programs):
    programs = list(programs)
    for move in moves:
        if move[0] == 's':
            programs = spin(int(move[1:]), programs)
        elif move[0] == 'x':
            a, b = map(int, move[1:].split('/'))
            programs = exchange(a, b, programs)
        elif move[0] == 'p':
            a, b = move[1:].split('/')
            programs = partner(a, b, programs)
    return ''.join(programs)

with open('input.txt', 'r') as file:
    moves = file.read().strip().split(',')

programs = 'abcdefghijklmnop'
print(dance(moves, programs))

# Part Two
programs = 'abcdefghijklmnop'
seen = []
for i in range(1000000000):
    if programs in seen:
        programs = seen[1000000000 % i]
        break
    seen.append(programs)
    programs = dance(moves, programs)

print(programs)
